### PR TITLE
Gaslib gas distribution model parsing tools

### PR DESCRIPTION
**Pull request by joshuahaddad**
## Fixes
NA

## Summary/Motivation:
This PR integrates work done to expand the capabilities in idaes.models_extra.gas_distribution. Included are two files that allow for the parsing and conversion of Gaslib network and scenario files to create idaes models.  

## Changes proposed in this PR:
- Addition of gaslib_parser which adds capabilities to read Gaslib XML files
- Addition of network_data which takes networks read by gaslib_parser and creates an idaes model for the network

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by ksbeattie** [comment]
@joshuahaddad I've put this on the Aug release and added a few reviewers.  Let me know if that's not correct.

File: idaes/models_extra/gas_distribution/parsers/tests/test_networkdata_model.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    data = urlopen("https://gaslib.zib.de/download/testData/GasLib-11.zip", timeout=3)
+    zipfile = ZipFile(BytesIO(data.read()))
+    files = [zipfile.open(file) for file in zipfile.namelist()]
+    netfile = files[1]
+    scnfile = files[2]
+    net = parse_gaslib_network(netfile, scnfile)
+    m = net.build_model(with_bounds=False)
+
+except timeout:
+    raise unittest.SkipTest('An internet connection is required to test gaslib parsing tools')
```
[the remark this answers is not in the document]
Comment (Robbybp) [reply idaes/models_extra/gas_distribution/parsers/tests/test_networkdata_model.py:39]: This also seems like useful functionality that we might want to import and use elsewhere. Maybe a function like the following could live in `gaslib_parser.py`?
```python
def get_gaslib_data(number=11, tout=3):
    if number != 11: 
        raise ValueError("Only GasLib-11 is supported for now.")
    url = "https://gaslib.zib.de/download/testData/GasLib-%s.zip" % int(number)
    try:
        data = urlopen(url, timeout=tout)
    except timeout:
        raise RuntimeError(
            "An internet connection is required to use gaslib parsing tools"
        )
    zipfile = ZipFile(BytesIO(data.read()))
    files = [zipfile.open(file) for file in zipfile.namelist()]
    netfiles = [f for f in files if f.name.endswith(".net")]
    scnfiles = [f for f in files if f.name.endswith(".scn")]
    if len(netfiles) > 1:
        raise RuntimeError(
            "More than one network file encountered when processing %s" % url 
        )
    if len(scnfiles) > 1:
        raise RuntimeError(
            "More than one scenario file encountered when processing %s" % url 
        )
    netfile = netfiles[0]
    scnfile = scnfiles[0]
    net = parse_gaslib_network(netfile, scnfile)
    return net 
```
Note that I tested this with the other GasLib test networks and got errors in the parser for all networks other than GasLib-11. Does this match your experience?

File: idaes/models_extra/gas_distribution/parsers/tests/test_networkdata_model.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+@pytest.mark.unit
```
Comment (Robbybp) [inline idaes/models_extra/gas_distribution/parsers/tests/test_networkdata_model.py:41]: Since these tests rely on data from an external source, they should probably be marked with `pytest.mark.integration` rather than unit.

File: idaes/models_extra/gas_distribution/parsers/tests/test_networkdata_model.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    # n5_x3 connection - pipe
+    n5_x3 = m.fs.pipelines['pipe08_N05_exit03']
+    assert(n5.outlets[1].arc.ports[0] == n5_x3.inlet_port)
+    assert(x3.inlets[0].arc.ports[0] == n5_x3.outlet_port)
+
+@pytest.mark.unit
+def test_nd_plot():
+    net.plot_results_at_time(m, m.fs.time.first(), show_plot=False)
+
+
```
Comment (Robbybp) [inline idaes/models_extra/gas_distribution/parsers/tests/test_networkdata_model.py:139]: Can you add a test that checks that the model has zero degrees of freedom? Maybe something like:
```python
from pyomo.contrib.incidence_analysis import IncidenceGraphInterface

def test_model_zero_dof():
    for node in m.fs.nodes.values():
        for supply in node.supplies.values():
            supply.state[:].pressure.fix()
    for comp in m.fs.compressors.values():
        comp.boost_pressure[:].fix()
    igraph = IncidenceGraphInterface(m, include_inequality=False)
    var_dmp, con_dmp = igraph.dulmage_mendelsohn()
    assert not var_dmp.unmatched and not con_dmp.unmatched
```
Alternatively you could use `assert not degrees_of_freedom(m)` where `degrees_of_freedom` comes from `idaes.core.util.model_statistics`, but I like Dulmage-Mendelsohn because this also makes sure the model is structurally nonsingular.

**Review by Robbybp — CHANGES_REQUESTED** [review]
Thanks for getting this PR up. I have some small change requests.

File: idaes/models_extra/gas_distribution/parsers/tests/test_networkdata_model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Robbybp) [reply idaes/models_extra/gas_distribution/parsers/tests/test_networkdata_model.py:39]: Note that the gaslib link here appears to be broken. The link that works for me now is:
```console
https://gaslib.zib.de/download/testData/GasLib-11-v1-20211130.zip
```
Maybe this is a good argument for not making this a publicly available function...

**Comment by ksbeattie** [comment]
@joshuahaddad, @Robbybp any update on this?  Moving to the November release board...

**Comment by joshuahaddad** [comment]
> @joshuahaddad, @Robbybp any update on this? Moving to the November release board...

Hey Keith, I left Sandia around this time and handed progress off to Michael. When I get home today I can submit the style changes when I get home today if needed. However, I don’t have a system that can run the code (needed a Linux optimizer build I believe).

File: idaes/models_extra/gas_distribution/parsers/tests/test_networkdata_model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (joshuahaddad) [reply idaes/models_extra/gas_distribution/parsers/tests/test_networkdata_model.py:39]: Hey Robbie sorry I didn’t see these comments before I left Sandia it was a bit hectic of a time… The code snippet for getting gaslib data seems to say only gaslib11 is supported. I believe Michael wrote this so I’d check with him to see if it’s been updated since.

**Comment by ksbeattie** [comment]
@Robbybp, @michaelbynum is this something you can take over?

**Comment by Robbybp** [comment]
@ksbeattie @michaelbynum I believe we can close this PR. These are useful tools that we would like to have if we were to continue working on pipeline models, but I don't believe anybody is doing this actively.

**Comment by michaelbynum** [comment]
I will pick this up in early October.

**Comment by andrewlee94** [comment]
@michaelbynum Any update on this PR? I am looking over our backlog of stale PRs to see what can/should be closed as inactive.

**Comment by michaelbynum** [comment]
I am going to get this done, but it is going to be a while. Let's close it until I can get back to it.
